yt/frontends/halo_catalog/data_structures.py

```diff
@@ -146,3 +177,363 @@ def _is_valid(self, *args, **kwargs):
             ):
                 return True
         return False
+
+
+class YTHaloParticleIndex(ParticleIndex):
+    """
+    Particle index for getting halo particles from YTHaloCatalogDatasets.
+    """
+
+    def __init__(self, ds, dataset_type):
+        self.real_ds = weakref.proxy(ds.real_ds)
+        super(YTHaloParticleIndex, self).__init__(ds, dataset_type)
+
+    def _calculate_particle_index_starts(self):
+        """
+        Create a dict of halo id offsets for each file.
+        """
+        particle_count = defaultdict(int)
+        offset_count = 0
+        for data_file in self.data_files:
+            data_file.index_start = dict(
+                [(ptype, particle_count[ptype]) for ptype in data_file.total_particles]
+            )
+            data_file.offset_start = offset_count
+            for ptype in data_file.total_particles:
+                particle_count[ptype] += data_file.total_particles[ptype]
+            offset_count += getattr(data_file, "total_offset", 0)
+
+        self._halo_index_start = dict(
+            [
+                (
+                    ptype,
+                    np.array(
+                        [data_file.index_start[ptype] for data_file in self.data_files]
+                    ),
+                )
+                for ptype in self.ds.particle_types_raw
+            ]
+        )
+
+    def _create_halo_id_table(self):
+        pass
```

Comment:
```suggestion
        raise NotImplementedError
```
not sure what the intention is here, might be the wrong approach but if that's the case then an inline comment would be nice for clarity :)